File: C Programming/DynamicArray/DynamicArrayDS.c
/*
    DynamicArrayDS.c
    - Dynamic array data structure & algorithm implementation.
*/
#include <stdio.h>
#include <stdlib.h>

int* mainArray = NULL;
int tailPointer = -1;
int mainArraySize = -1;

int getTailPointer() {
    return tailPointer;
}

void initMainArray(int inputLength) {
    if (inputLength <= 0) {
        printf("ERROR: cannot initialize mainArray[] with inputLength = %d.\n", inputLength);
    } else {
        printf("SUCCESS: initializing mainArray[] to length = %d.\n", inputLength);
        mainArraySize = inputLength;
        mainArray = (int*) malloc(mainArraySize * sizeof(int));
        tailPointer++;
    }
}

int* shiftRight(int* inputArray, int tailPointer) {
    if (tailPointer == 1) {
        inputArray[1] = inputArray[0];
        inputArray[0] = 0;
        return inputArray;
    } else {
        for (int i = tailPointer; i >= 1; i--) {
            inputArray[i] = inputArray[i - 1];
        }
        return inputArray;
    }
}

int* shiftLeft(int* inputArray, int tailPointer) {
    for (int i = 0; i < tailPointer - 1; i++) {
        inputArray[i] = inputArray[i + 1];
    }
    inputArray[tailPointer - 1] = 0;
    return inputArray;
}

void insertAtHead(int data) {
    if (mainArray == NULL  || tailPointer == -1) {
        printf("\nERROR: cannot insertAtHead(%d) in uninitialized/NULL mainArray[].\n", data);
    } else {
        if (tailPointer == mainArraySize) {
            mainArraySize++;
            printf("\nResizing mainArray[] to length/size = %d.\n", mainArraySize);
            printf("SUCCESS: insertingAtHead(%d) in mainArray[].\n", data);
            int* tempArray = (int*) malloc(mainArraySize * sizeof(int));
            for (int i = 0; i < mainArraySize - 1; i++) {
                tempArray[i + 1] = mainArray[i];
            }
            tempArray[0] = data;
            mainArray = tempArray;
            tailPointer++;
        } else {
            printf("\nSUCCESS: insertingAtHead(%d) in mainArray[].\n", data);
            mainArray = shiftRight(mainArray, tailPointer);
            mainArray[0] = data;
            tailPointer++;
        }
    }
}

void removeFromHead() {
    if (mainArray == NULL || tailPointer == -1) {
        printf("\nERROR: cannot removeFromHead() from uninitialized/NULL mainArray[].\n");
    } else {
        if (tailPointer == 0) {
            printf("\nERROR: cannot removeFromHead() from empty mainArray[].\n");
        } else {
            printf("\nSUCCESS: removingFromHead(%d) from mainArray[].\n", mainArray[0]);
            mainArray = shiftLeft(mainArray, tailPointer);
            tailPointer--;
        }
    }
}

void insertAtTail(int data) {
    if (mainArray == NULL || tailPointer == -1) {
        printf("\nERROR: cannot insertAtTail(%d) in uninitialized/NULL mainArray[].\n", data);
    } else {
        if (tailPointer == mainArraySize) {
            mainArraySize++;
            printf("\nResizing mainArray[] to length/size = %d.\n", mainArraySize);
            printf("SUCCESS: insertingAtTail(%d) in mainArray[].\n", data);
            int* tempArray = (int*) malloc(mainArraySize * sizeof(int));
            for (int i = 0; i < mainArraySize - 1; i++) {
                tempArray[i] = mainArray[i];
            }
            tempArray[tailPointer] = data;
            mainArray = tempArray;
            tailPointer++;
        } else {
            printf("\nSUCCESS: insertingAtTail(%d) in mainArray[].\n", data);
            mainArray[tailPointer] = data;
            tailPointer++;
        }
    }
}

void removeFromTail() {
    if (mainArray == NULL || tailPointer == -1) {
        printf("\nERROR: cannot removeFromTail() from uninitialized/NULL mainArray[].\n");
    } else {
        if (tailPointer == 0) {
            printf("\nERROR: cannot removeFromTail() from empty mainArray[].\n");
            return;
        }
        printf("\nSUCCESS: removingFromTail(%d) from mainArray[].\n", mainArray[tailPointer - 1]);
        mainArray[tailPointer - 1] = 0;
        tailPointer--;
    }
}

void printArray() {
    if (mainArray == NULL || tailPointer == -1) {
        printf("mainArray[]: []\n");
    } else {
        printf("mainArray[]: [");
        for (int i = 0; i < mainArraySize; i++) {
            printf("%d, ", mainArray[i]);
        }
        printf("\b\b]\n");
    }
}

int main(int argc, char* argv[]) {
    printf("StaticArrayDS.c\n");
    printf("---------------\n");

    // printf("\narr1 -->\n");
    // initMainArray(0);
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // insertAtHead(3);
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // removeFromHead();
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // insertAtTail(22);
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // removeFromTail();
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());

    // printf("\narr2 -->\n");
    // initMainArray(1);
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // insertAtHead(3);
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // removeFromHead();
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // insertAtTail(22);
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());
    // removeFromTail();
    // printArray();
    // printf("tailPointer = %d\n", getTailPointer());

    printf("\narr3 -->\n");
    initMainArray(5);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtHead(3);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtTail(22);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtHead(7);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtTail(81);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    removeFromHead();
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    removeFromTail();
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    removeFromHead();
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    removeFromTail();
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    removeFromHead();
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtHead(200);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtTail(300);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtHead(100);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtTail(400);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtHead(75);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtTail(500);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());
    insertAtTail(600);
    printArray();
    printf("tailPointer = %d\n", getTailPointer());

    return EXIT_SUCCESS;
}

File: Java Programming/DynamicArray/DynamicArrayDS.java
/*
    DynamicArrayDS.java
    - Dynamic array data structure & algorithm implementation.
*/

public class DynamicArrayDS {
    int[] mainArray = null;
    int tailPointer = -1;

    public DynamicArrayDS(int inputLength) {
        if (inputLength <= 0) {
            System.out.println("ERROR: Cannot initialize mainArray[] with inputLength = " + inputLength + ".");
        } else {
            System.out.println("SUCCESS: Initializing mainArray[] to length = " + inputLength + ".");
            mainArray = new int[inputLength];
            tailPointer++;
        }
    }

    public int getTailPointer() {
        return this.tailPointer;
    }

    public void insertAtHead(int data) {
        if (mainArray == null || tailPointer == -1) {
            System.out.println("\nERROR: Cannot insertAtHead(" + data + ") in uninitialized/null mainArray[].");
        } else {
            if (tailPointer == mainArray.length) {
                System.out.println("\nResizing mainArray[] to length/size = " + (mainArray.length + 1) + ".");
                System.out.println("SUCCESS: insertingAtHead(" + data + ") in mainArray[].");
                int newLength = mainArray.length + 1;
                int[] tempArray = new int[newLength];
                for (int i = 0; i < tempArray.length - 1; i++) {
                    tempArray[i + 1] = mainArray[i];
                }
                tempArray[0] = data;
                mainArray = tempArray;
                tailPointer++;
            } else {
                System.out.println("\nSUCCESS: insertingAtHead(" + data + ") in mainArray[].");
                mainArray = shiftRight(mainArray, tailPointer);
                mainArray[0] = data;
                tailPointer++;
            }
        }
    }

    public void removeFromHead() {
        if (mainArray == null || tailPointer == -1) {
            System.out.println("\nERROR: Cannot removeFromHead() from uninitialized/null mainArray[].");
        } else {
            if (tailPointer == 0) {
                System.out.println("\nERROR: Cannot removeFromHead() from empty mainArray[].");
                return;
            }
            System.out.println("\nSUCCESS: removingFromHead(" + mainArray[0] + ") from mainArray[].");
            shiftLeft(mainArray, tailPointer);
            tailPointer--;
        }
    }

    public void insertAtTail(int data) {
        if (mainArray == null || tailPointer == -1) {
            System.out.println("\nERROR: Cannot insertAtTail(" + data + ") in uninitialized/null mainArray[].");
        } else {
            if (tailPointer == mainArray.length) {
                System.out.println("\nResizing mainArray[] to length/size = " + (mainArray.length + 1) + ".");
                System.out.println("SUCCESS: insertingAtHead(" + data + ") in mainArray[].");
                int newLength = mainArray.length + 1;
                int[] tempArray = new int[newLength];
                for (int i = 0; i < newLength - 1; i++) {
                    tempArray[i] = mainArray[i];
                }
                tempArray[tailPointer] = data;
                mainArray = tempArray;
                tailPointer++;
            } else {
                System.out.println("\nSUCCESS: insertingAtTail(" + data + ") in mainArray[].");
                mainArray[tailPointer] = data;
                tailPointer++;
            }
        }
    }

    public void removeFromTail() {
        if (mainArray == null || tailPointer == -1) {
            System.out.println("\nERROR: Cannot removeFromTail() from uninitialized/null mainArray[].");
        } else {
            if (tailPointer == 0) {
                System.out.println("\nERROR: Cannot removeFromTail() from empty mainArray[].");
                return;
            }
            System.out.println("\nSUCCESS: removingFromTail(" + mainArray[tailPointer - 1] + ") from mainArray[].");
            mainArray[tailPointer - 1] = 0;
            tailPointer--;
        }
    }

    private int[] shiftRight(int[] inputArray, int tailPointer) {
        if (tailPointer == 1) {
            inputArray[1] = inputArray[0];
            inputArray[0] = 0;
            return inputArray;
        } else {
            for (int i = tailPointer; i >= 1; i--) {
                inputArray[i] = inputArray[i - 1];
            }
            return inputArray;
        }
    }

    private int[] shiftLeft(int[] inputArray, int tailPointer) {
        for (int i = 0; i < tailPointer - 1; i++) {
            inputArray[i] = inputArray[i + 1];
        }
        inputArray[tailPointer - 1] = 0;
        return inputArray;
    }

    public void printArray() {
        if (mainArray == null || tailPointer == -1) {
            System.out.println("mainArray[]: []");
        } else {
            System.out.print("mainArray[]: [");
            for (int i = 0; i < mainArray.length; i++) {
                System.out.print(mainArray[i] + ", ");
            }
            System.out.print("\b\b]\n");
        }
    }
}